src/MemStore.cc

```diff
@@ -569,7 +569,7 @@ MemStore::copyFromShmSlice(StoreEntry &e, const StoreIOBuffer &buf, bool eof)
 
     // local memory stores both headers and body so copy regardless of pstate
     const int64_t offBefore = e.mem_obj->endOffset();
-    assert(e.mem_obj->data_hdr.write(buf)); // from MemObject::write()
+    e.writeData(buf);
     const int64_t offAfter = e.mem_obj->endOffset();
```

Comment:
I am thinking about a different approach that would allow us to remove this case without resorting to this 'early' locking approach. What if we convert our mem_hdr::isIdle from bool to Optional which will get its initial value just after becoming idle, i.e., StoreEntry::unlock() setting lock_count to zero? In this approach, pages written to the newly created and still unlocked entry (e.g, being copied from shared memory) are not 'idle' by definition and should not be taken into account.

Reply:
Introducing a third state (busy, idle, _initial_) feels like a rather heavy workaround that makes things worse for everybody long-term in the name of accommodating one special case that is known to be conceptually broken.

AFAICT, an ideal solution would follow these simple and strict rules[^1]:

* lock count increases from 0 to 1: Move mem_hdr pages from the idle pool into the busy pool
* lock count decreases from 1 to 0: Move mem_hdr pages from the busy pool into the idle pool
* new page added: assert that the entry is locked; add the page to the busy pool
* old page removed: assert that the entry is locked; remove the page from the busy pool

We cannot afford the above level of simplicity and strictness because MemStore-related code and StoreEntry destruction code (at least) modify an unlocked entry. We have two valid/good options AFAICT: Fix known problem(s) or stop enforcing higher-level principles in low-level mem_hdr code (while still counting everything correctly). If the former is too difficult, the latter leads to the adjustment in the last two rules:

* lock count increases from 0 to 1: Move mem_hdr pages from the idle pool into the busy pool
* lock count decreases from 1 to 0: Move mem_hdr pages from the busy pool into the idle pool
* new page added: Add it to the current[^2] pool 
* old page removed: Remove it from the current[^2] pool

The last two rules require either storing the "current pool" state in mem_hdr::isIdle or passing it to every add/remove page operation. Either is better than introducing a third pooling state IMO (imagine what the equivalent 5(?) rules would look like if we add that third state)!

[^1]: We do not have such actual _pools_ or even _two_ counters right now IIRC, but, conceptually, we can think of the problem in terms of those busy and idle page pools. The difference is just an implementation detail in this discussion context AFAICT.

[^2]: The busy pool (if the entry is locked) or the idle pool (otherwise).

Reply:
> However, an entry may be unlocked and still written when requesting it, e.g., from the shared memory store.

I think the shared memory case should be excluded because if it is on, StoreEntry is not placed into a memory policy (StoreEntry::setMemStatus()). We have also store_client::readBody(), but it is called after Controller::find() returns the found entry and the caller locks it (however, nobody probably forbid to call it inside Controller::find()).

I adjusted the code, something along with your second set of rules (2340315).